plugins/Utils/windowstatestorage.cpp

```diff
@@ -16,167 +17,329 @@
-{
-    const QString queryString = QStringLiteral("SELECT stage FROM stage WHERE appId = '%1';")
-            .arg(sanitiseString(appId));
+    void saveStage(const QString &appId, int stage)
+    {
+        QSqlDatabase connection = QSqlDatabase::database(m_connectionName);
+        QSqlQuery query(connection);
+        query.prepare(m_saveStageQuery);
+        query.bindValue(":appId", appId);
+        query.bindValue(":stage", stage);
+        if (!query.exec()) {
+            logSqlError(query);
+        }
+    }
 
-    QSqlQuery query = getValue(queryString);
+private:
+    const QString m_connectionName = QStringLiteral("WindowStateStorage");
+    const QString m_getStateQuery = QStringLiteral("SELECT state FROM state WHERE windowId = :windowId");
+    const QString m_saveStateQuery = QStringLiteral("INSERT OR REPLACE INTO state (windowId, state) values (:windowId, :state)");
+    const QString m_getGeometryQuery = QStringLiteral("SELECT * FROM geometry WHERE windowId = :windowId");
+    const QString m_saveGeometryQuery = QStringLiteral("INSERT OR REPLACE INTO geometry (windowId, x, y, width, height) values (:windowId, :x, :y, :width, :height)");
+    const QString m_getStageQuery = QStringLiteral("SELECT stage FROM stage WHERE appId = :appId");
+    const QString m_saveStageQuery = QStringLiteral("INSERT OR REPLACE INTO stage (appId, stage) values (:appId, :stage)");
+    QString m_dbName;
+
+    void logSqlError(const QSqlQuery query) const
+    {
+        qWarning() << "Error executing query" << query.lastQuery()
+                << "Driver error:" << query.lastError().driverText()
+                << "Database error:" << query.lastError().databaseText();
+    }
+};
 
-    if (!query.first()) {
-        return defaultValue;
+WindowStateStorage::WindowStateStorage(const QString& dbName, QObject *parent):
+    QObject(parent),
+    m_thread()
+{
+    qRegisterMetaType<WindowStateStorage::WindowState>("WindowStateStorage::WindowState");
+    QString dbFile;
+    if (dbName == nullptr) {
+        const QString dbPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + QStringLiteral("/unity8/");
+        QDir dir;
+        dir.mkpath(dbPath);
+        dbFile = QString(dbPath) + QStringLiteral("windowstatestorage.sqlite");
+    } else {
+        dbFile = dbName;
+    }
+    m_asyncQuery = new AsyncQuery(dbFile);
+    m_asyncQuery->moveToThread(&m_thread);
+    connect(&m_thread, &QThread::finished, m_asyncQuery, &QObject::deleteLater);
+    m_thread.start();
+    bool queryInitSuccessful;
+    QMetaObject::invokeMethod(m_asyncQuery, "initdb",
+                              Qt::BlockingQueuedConnection,
+                              Q_RETURN_ARG(bool, queryInitSuccessful));
+    if (!queryInitSuccessful) {
+        qWarning() << "WindowStateStorage Failed to initialize AsyncQuery! Windows will not be restored to their previous location.";
+        m_asyncOk = false;
+        m_thread.quit();
```

Comment:
We may want to nullify `m_asyncQuery` here (and then `m_asyncOk` isn't really needed), unless you decided to implement @mariogrip's comment at line 259 (see my comment).